# The goal of this EDA is to explore and understand the IBM HR Analytics dataset, identify key factors influencing employee attrition, and uncover patterns in the data through visualizations and statistical summaries. 

_Importing the necessary libraries required for data loading, analysis, and visualization._

In [37]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression

_Loading the dataset into a pandas DataFrame_

# Data Loading

In [38]:
df = pd.read_csv("data/WA_Fn-UseC_-HR-Employee-Attrition.csv")

_Taking a quick look at the dataset to understand its structure, data types, and sample records_

In [39]:
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

_Checking the shape of the dataset to see how many rows (records) and columns (features) it contains_

In [40]:
df.shape

(1470, 35)

_The dataset contains 1,470 rows and 35 columns, meaning we have information on 1,470 employees and 35 different features related to their personal and professional details_

_Getting a summary of the dataset, including column names, data types, and non-null value counts_

In [41]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

_The dataset consists of 26 numerical features and 9 categorical features_

# Data Preprocessing

_Taking a quick look at thenumerical features_

In [42]:
df.select_dtypes(include='int64')

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
0      41       1102                 1          2              1   
1      49        279                 8          1              1   
2      37       1373                 2          2              1   
3      33       1392                 3          4              1   
4      27        591                 2          1              1   
...   ...        ...               ...        ...            ...   
1465   36        884                23          2              1   
1466   39        613                 6          1              1   
1467   27        155                 4          3              1   
1468   49       1023                 2          3              1   
1469   34        628                 8          3              1   

      EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
0                  1                        2          94               3   
1                  2                        3          61               2   
2                  4                        4          92               2   
3                  5                        4          56               3   
4                  7                        1          40               3   
...              ...                      ...         ...             ...   
1465            2061                        3          41               4   
1466            2062                        4          42               2   
1467            2064                        2          87               4   
1468            2065                        4          63               2   
1469            2068                        2          82               4   

      JobLevel  ...  RelationshipSatisfaction  StandardHours  \
0            2  ...                         1             80   
1            2  ...                         4             80   
2            1  ...                         2             80   
3            1  ...                         3             80   
4            1  ...                         4             80   
...        ...  ...                       ...            ...   
1465         2  ...                         3             80   
1466         3  ...                         1             80   
1467         2  ...                         2             80   
1468         2  ...                         4             80   
1469         2  ...                         1             80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

      WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                   1               6                   4   
1                   3              10                   7   
2                   3               0                   0   
3                   3               8                   7   
4                   3               2                   2   
...               ...             ...                 ...   
1465                3               5                   2   
1466                3               7                   7   
1467                3               6                   2   
1468                2               9                   6   
1469   

_Some of the numerical columns appear to be ordinal categorical features that store numeric values representing categories. Let's find them out_

In [43]:
df.select_dtypes(include='int64').nunique()

Age                           43
DailyRate                    886
DistanceFromHome              29
Education                      5
EmployeeCount                  1
EmployeeNumber              1470
EnvironmentSatisfaction        4
HourlyRate                    71
JobInvolvement                 4
JobLevel                       5
JobSatisfaction                4
MonthlyIncome               1349
MonthlyRate                 1427
NumCompaniesWorked            10
PercentSalaryHike             15
PerformanceRating              2
RelationshipSatisfaction       4
StandardHours                  1
StockOptionLevel               4
TotalWorkingYears             40
TrainingTimesLastYear          7
WorkLifeBalance                4
YearsAtCompany                37
YearsInCurrentRole            19
YearsSinceLastPromotion       16
YearsWithCurrManager          18
dtype: int64

_For ease of analysis and visualization, we'll convert the ordinal categorical columns stored as numbers into their corresponding labeled categories, making the data more interpretable_

In [44]:
df["Education"] = df["Education"].replace({1:"Below College",2:"College",3:"Bachelor",4:"Master",5:"Doctor"})
df["EnvironmentSatisfaction"] = df["EnvironmentSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})
df["JobInvolvement"] = df["JobInvolvement"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})
df["JobLevel"] = df["JobLevel"].replace({1:"Entry Level",2:"Junior Level",3:"Mid Level",4:"Senior Level",5:"Executive Level"})
df["JobSatisfaction"] = df["JobSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})
df["PerformanceRating"] = df["PerformanceRating"].replace({1:"Low",2:"Good",3:"Excellent",4:"Outstanding"})
df["RelationshipSatisfaction"] = df["RelationshipSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})
df["WorkLifeBalance"] = df["WorkLifeBalance"].replace({1:"Bad",2:"Good",3:"Better",4:"Best"})

_We observe that the EmployeeCount column has only one unique value (1), indicating that each row represents a single employee, but it does not add any analytical value, so it can be dropped. Similarly, the StandardHours column contains a constant value of 80 for all records, making it irrelevant for our analysis. Additionally, the EmployeeNumber column is a unique identifier for each employee and does not contribute any meaningful insight toward attrition prediction. Therefore, all three columns will be removed from the dataset to keep only relevant features_

In [45]:
df.drop(['EmployeeCount', 'EmployeeNumber', 'StandardHours'], axis="columns", inplace=True)

_Let’s take another look at the numerical features after removing the irrelevant columns_

In [46]:
df.select_dtypes(include='int64')

Age  DailyRate  DistanceFromHome  HourlyRate  MonthlyIncome  \
0      41       1102                 1          94           5993   
1      49        279                 8          61           5130   
2      37       1373                 2          92           2090   
3      33       1392                 3          56           2909   
4      27        591                 2          40           3468   
...   ...        ...               ...         ...            ...   
1465   36        884                23          41           2571   
1466   39        613                 6          42           9991   
1467   27        155                 4          87           6142   
1468   49       1023                 2          63           5390   
1469   34        628                 8          82           4404   

      MonthlyRate  NumCompaniesWorked  PercentSalaryHike  StockOptionLevel  \
0           19479                   8                 11                 0   
1           24907                   1                 23                 1   
2            2396                   6                 15                 0   
3           23159                   1                 11                 0   
4           16632                   9                 12                 1   
...           ...                 ...                ...               ...   
1465        12290                   4                 17                 1   
1466        21457                   4                 15                 1   
1467         5174                   1                 20                 1   
1468        13243                   2                 14                 0   
1469        10228                   2                 12                 0   

      TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                     8                      0               6   
1                    10                      3              10   
2                     7                      3               0   
3                     8                      3               8   
4                     6                      3               2   
...                 ...                    ...             ...   
1465                 17                      3               5   
1466                  9                      5               7   
1467                  6                      0               6   
1468                 17                      3               9   
1469                  6                      3               4   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                      4                        0                     5  
1                      7                        1                     7  
2                      0                        0                     0  
3                      7                        3                     0  
4                      2                        2                     2  
...                  ...                      ...                   ...  
1465                   2                        0                     3  
1466                   7                        1                     7  
1467                   2                        0                     3  
1468                   6                        0                     8  
1469                   3                        1                     2  

[1470 rows x 15 columns]

_After cleaning the dataset and removing irrelevant columns, we are left with 15 numerical features that will be used for analysis_

_Now, let's clean the categorical features_

In [47]:
df.select_dtypes(include='object')

Attrition     BusinessTravel              Department      Education  \
0          Yes      Travel_Rarely                   Sales        College   
1           No  Travel_Frequently  Research & Development  Below College   
2          Yes      Travel_Rarely  Research & Development        College   
3           No  Travel_Frequently  Research & Development         Master   
4           No      Travel_Rarely  Research & Development  Below College   
...        ...                ...                     ...            ...   
1465        No  Travel_Frequently  Research & Development        College   
1466        No      Travel_Rarely  Research & Development  Below College   
1467        No      Travel_Rarely  Research & Development       Bachelor   
1468        No  Travel_Frequently                   Sales       Bachelor   
1469        No      Travel_Rarely  Research & Development       Bachelor   

     EducationField EnvironmentSatisfaction  Gender JobInvolvement  \
0     Life Sciences                  Medium  Female           High   
1     Life Sciences                    High    Male         Medium   
2             Other               Very High    Male         Medium   
3     Life Sciences               Very High  Female           High   
4           Medical                     Low    Male           High   
...             ...                     ...     ...            ...   
1465        Medical                    High    Male      Very High   
1466        Medical               Very High    Male         Medium   
1467  Life Sciences                  Medium    Male      Very High   
1468        Medical               Very High    Male         Medium   
1469        Medical                  Medium    Male      Very High   

          JobLevel                    JobRole JobSatisfaction MaritalStatus  \
0     Junior Level            Sales Executive       Very High        Single   
1     Junior Level         Research Scientist          Medium       Married   
2      Entry Level      Laboratory Technician            High        Single   
3      Entry Level         Research Scientist            High       Married   
4      Entry Level      Laboratory Technician          Medium       Married   
...            ...                        ...             ...           ...   
1465  Junior Level      Laboratory Technician       Very High       Married   
1466     Mid Level  Healthcare Representative             Low       Married   
1467  Junior Level     Manufacturing Director          Medium       Married   
1468  Junior Level            Sales Executive          Medium       Married   
1469  Junior Level      Laboratory Technician            High       Married   

     Over18 OverTime PerformanceRating RelationshipSatisfaction  \
0         Y      Yes         Excellent                      Low   
1         Y       No       Outstanding                Very High   
2         Y      Yes         Excellent                   Medium   
3         Y      Yes         Excellent                     High   
4         Y       No         Excellent                Very High   
...     ...      ...               ...                      ...   
1465      Y       No         Excellent                     High   
1466      Y       No         Excellent                      Low   
1467      Y      Yes       Outstanding                   Medium   
1468      Y       No         Excellent                Very High   
1469      Y       No         Excellent                      Low   

     WorkLifeBalance  
0                Bad  
1             Better  
2             Better  
3             Better  
4             Better  
...              ...  
1465          Better  
1466          Better  
1467          Better  
1468            Good  
1469            Best  

[1470 rows x 17 columns]

In [48]:
df.select_dtypes(include='object').nunique()

Attrition                   2
BusinessTravel              3
Department                  3
Education                   5
EducationField              6
EnvironmentSatisfaction     4
Gender                      2
JobInvolvement              4
JobLevel                    5
JobRole                     9
JobSatisfaction             4
MaritalStatus               3
Over18                      1
OverTime                    2
PerformanceRating           2
RelationshipSatisfaction    4
WorkLifeBalance             4
dtype: int64

_The minimum age in the dataset is 18, which confirms that all employees are adults. Therefore, the Over18 attribute is not need. So, we will drop it as well_

In [49]:
df.drop(['Over18'], axis="columns", inplace=True)

In [50]:
df.select_dtypes(include='object').shape

(1470, 16)

_Now, we have 16 Categorical features_

_Let’s quickly glance over the cleaned dataset_

In [51]:
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome      Education EducationField EnvironmentSatisfaction  \
0                 1        College  Life Sciences                  Medium   
1                 8  Below College  Life Sciences                    High   
2                 2        College          Other               Very High   
3                 3         Master  Life Sciences               Very High   
4                 2  Below College        Medical                     Low   

   Gender  ...  PerformanceRating RelationshipSatisfaction StockOptionLevel  \
0  Female  ...          Excellent                      Low                0   
1    Male  ...        Outstanding                Very High                1   
2    Male  ...          Excellent                   Medium                0   
3  Female  ...          Excellent                     High                0   
4    Male  ...          Excellent                Very High                1   

  TotalWorkingYears TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                 8                     0             Bad               6   
1                10                     3          Better              10   
2                 7                     3          Better               0   
3                 8                     3          Better               8   
4                 6                     3          Better               2   

   YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
0                   4                        0                    5  
1                   7                        1                    7  
2                   0                        0                    0  
3                   7                        3                    0  
4                   2                        2                    2  

[5 rows x 31 columns]

In [52]:
df.shape

(1470, 31)

_we are now left with 31 meaningful features in the dataset_

# Exploratory Data Analysis